src/pages/index.js

```diff
@@ -1,93 +1,225 @@
-    },
+const api = new Api({
+  baseUrl: "https://around-api.en.tripleten-services.com/v1",
+  headers: {
+    authorization: "9aa267fe-3315-41ab-a072-2d6dbd5ca77b",
+    "Content-Type": "application/json",
   },
-  selectors.cardList
-);
-
-const userInfoPopup = new PopupWithForm(selectors.editFormPopup, (data) => {
-  userInfo.setUserInfo(data);
 });
 
-const newCardPopup = new PopupWithForm(selectors.addFormPopup, (data) => {
-  const card = createCard(data);
-  cardSection.addItem(card);
-});
+let cardSection;
+
+api
+  .getAppInfo()
+  .then(([initialCards, userData]) => {
+    userInfo.setUserInfo(userData.name, userData.about);
+    userInfo.setAvatar(userData.avatar);
+    cardSection = new Section(
+      {
+        items: initialCards.reverse(),
+        renderer: (data) => {
+          const card = renderCard(data);
+          cardSection.addItem(card);
+        },
+      },
+      selectors.cardList
+    );
+    cardSection.renderItems();
+  })
+  .catch((err) => {
+    console.error(err);
+  });
+
+// Validation
 
-// VALIDATION //
 const formValidators = {};
 
 const enableValidation = (config) => {
   const formList = Array.from(document.querySelectorAll(config.formSelector));
   formList.forEach((formElement) => {
     const validator = new FormValidator(config, formElement);
     const formName = formElement.getAttribute("name");
-
-    formValidators[formName] = validator;
-    validator.enableValidation();
+    if (formName) {
+      formValidators[formName] = validator;
+      validator.enableValidation();
+    }
   });
 };
 
-//Event Listeners
-profileEditButton.addEventListener("click", () => {
-  fillProfileForm();
-  formValidators[selectors.profileForm].resetFormValidation();
-  userInfoPopup.open();
-});
+// Create Instances
 
-addNewCardButton.addEventListener("click", () => {
-  newCardPopup.open();
+const cardPreviewPopup = new PopupWithImage(selectors.previewPopup);
+
+const userInfo = new UserInfo(
+  selectors.profileTitle,
+  selectors.profileParagraph,
+  selectors.profileImage
+);
+
+const deleteCardPopup = new PopupWithConfirm(selectors.deleteCardPopup);
+
+const addCardPopup = new PopupWithForm(selectors.addFormPopup, (cardData) => {
+  addCardPopup.setLoading(true);
+  api
+    .createNewCard(cardData)
+    .then((res) => {
+      const card = renderCard(res);
+      cardSection.addItem(card);
+      addCardPopup.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      addCardPopup.setLoading(false);
+    });
 });
 
+const ProfileEditPopup = new PopupWithForm(
+  selectors.editFormPopup,
+  (userData) => {
+    ProfileEditPopup.setLoading(true);
+    api
+      .updateUserInfo(userData)
+      .then(() => {
+        userInfo.setUserInfo(userData.name, userData.about);
+        ProfileEditPopup.close();
+      })
+      .catch((err) => {
+        console.error(err);
+      })
+      .finally(() => {
+        ProfileEditPopup.setLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

Reply:
This was a little bit hard to understand, but i did it and i like this approach. Thank you!